### PR TITLE
all: EIP-4844 integration 

### PR DESCRIPTION
**Pull request by MariusVanDerWijden**
4844 integration PR

### CODE REVIEW

File: core/types/tx_blob.go [hunk 1]
```diff
@@ -20,28 +20,48 @@ import (
...
 	// Signature values
-	V *uint256.Int `json:"v" gencodec:"required"`
-	R *uint256.Int `json:"r" gencodec:"required"`
-	S *uint256.Int `json:"s" gencodec:"required"`
+	V *big.Int `json:"v" gencodec:"required"`
+	R *big.Int `json:"r" gencodec:"required"`
+	S *big.Int `json:"s" gencodec:"required"`
+}
+
+type BlobTxWithBlobs struct {
```
Comment (karalabe) [inline core/types/tx_blob.go:47]: This should not exist in the consensus types at all. This is a network level thing, pls move it to eth/protocols/eth

File: core/state_transition.go [hunk 2]
```diff
@@ -230,12 +233,30 @@ func (st *StateTransition) to() common.Address {
...
+				panic("adsf")
+			}
+			// Check that the user has enough funds to cover dataGasUsed * tx.BlobGasFeeCap
+			blobBalanceCheck := new(big.Int).SetUint64(dataGas)
+			blobBalanceCheck.Mul(blobBalanceCheck, st.msg.BlobGasFeeCap)
+			balanceCheck.Add(balanceCheck, blobBalanceCheck)
+			// Pay for dataGasUsed * actual blob fee
+			blobFee := new(big.Int).SetUint64(dataGas)
+			blobFee.Mul(blobFee, eip4844.CalcBlobFee(*st.evm.Context.ExcessDataGas))
+			mgval.Add(mgval, blobFee)
```
Comment (holiman) [inline core/state_transition.go:256]: this seems dangerous. See up above? 
```
balanceCheck := mgval
```
So at this point, the one might alias the other. So IMO let's change that into `balanceCheck := new(big.Int).Set(mgval)` before this clause, so we don't change one when we want to change the other. Sneaky bugs

File: consensus/beacon/consensus.go [hunk 3]
```diff
@@ -372,6 +373,14 @@ func (beacon *Beacon) FinalizeAndAssemble(chain consensus.ChainHeaderReader, hea
...
 		}
 	}
+	if chain.Config().IsCancun(header.Number, header.Time) {
+		var blobs int
+		for _, tx := range txs {
+			blobs += len(tx.BlobHashes())
+		}
+		dataGasUsed := uint64(blobs * params.BlobTxDataGasPerBlob)
+		header.DataGasUsed = &dataGasUsed
+	}
```
Comment (fjl) [inline consensus/beacon/consensus.go:383]: Is this really the best place to update the header? It kind of seems like this belongs into the state transition, where we usually track the block gas?

File: miner/payload_building.go [hunk 4]
```diff
@@ -148,15 +155,15 @@ func (payload *Payload) ResolveFull() *engine.ExecutionPayloadEnvelope {
...
+	return engine.BlockToExecutableData(payload.full, payload.fullFees, payload.blobs, payload.commitments, payload.proofs)
 }
 
 // buildPayload builds the payload according to the provided parameters.
 func (w *worker) buildPayload(args *BuildPayloadArgs) (*Payload, error) {
 	// Build the initial version with no transaction included. It should be fast
 	// enough to run. The empty payload can at least make sure there is something
 	// to deliver for not missing slot.
-	empty, _, err := w.getSealingBlock(args.Parent, args.Timestamp, args.FeeRecipient, args.Random, args.Withdrawals, true)
+	empty, _, _, _, _, err := w.getSealingBlock(args.Parent, args.Timestamp, args.FeeRecipient, args.Random, args.Withdrawals, true)
```
Comment (fjl) [inline miner/payload_building.go:166]: Dog sled situation here.

**Comment by fjl** [comment]
We have this in now, and all fixes for Cancun should go to master branch.

**Pull request closed without merging**